Add tests

jat-canonical opened the pull request:
# Done 

* Add basic tests to subcluster and main modules
* Add the portion to run these tests in CI

# Jira Tickets

Fixes [AC-3144](https://warthogs.atlassian.net/browse/AC-3144)
Fixes [AC-3145](https://warthogs.atlassian.net/browse/AC-3145)

[AC-3144]: https://warthogs.atlassian.net/browse/AC-3144?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[AC-3145]: https://warthogs.atlassian.net/browse/AC-3145?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

jat-canonical committed:
tests(subcluster): add terraform tests to subcluster

This commit uses `terraform test`, which is the inbuilt way to inspect
terraform plans, to assert certain defaults and model layout for the
anbox subcluster.

jat-canonical committed:
tests(main): add basic unit tests to the actual plan

jat-canonical committed:
feat(gh): add job to run terraform tests

jat-canonical committed:
feat: Add anbox team as codeowners

morphis commented on .github/workflows/build-and-test.yaml:
```diff
@@ -32,6 +32,27 @@ jobs:
+        with:
+          terraform_version: 1.11.2
+      - name: Setup operator environment
+        uses: charmed-kubernetes/actions-operator@1c7c9a30d7d233e26e7a4fc1505cc44bbd937229
```
I see why but I would like to reduce the number of 3rd party actions to a minimum. From what I see we should be able to do the same with `canonical/setup-lxd` and a directly install of `juju` and doing the bootstrap, right?

jat-canonical replied:
I though this is also maintained by us only ? I see [this ](https://github.com/charmed-kubernetes/actions-operator/network/dependents?dependents_before=NDI5MjcwMjg3NTk) list and majority of the repos are canonical owned ?

morphis commented on .github/workflows/build-and-test.yaml:
```diff
@@ -32,6 +32,27 @@ jobs:
+      - name: Checkout
+        uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4
+      - name: Setup Terraform
+        uses: hashicorp/setup-terraform@b9cd54a3c349d3f38e8881555d616ced269862dd # v3
```
Shouldn't we use the terraform snap?

morphis commented on modules/subcluster/tests/base_deploy.tftest.hcl:
```diff
@@ -0,0 +1,124 @@
+run "test_required_variables" {
```
Can we have a copyright header?

morphis commented on modules/subcluster/tests/dashboard.tftest.hcl:
```diff
@@ -0,0 +1,91 @@
+run "test_dashboard_deploy_only_when_enabled_with_streaming_stack" {
```
Same here.

morphis commented on modules/subcluster/tests/external_etcd.tftest.hcl:
```diff
@@ -0,0 +1,32 @@
+run "test_external_etcd_enabled" {
```
Same here.

morphis commented on tests/basic_subcluster.tftest.hcl:
```diff
@@ -0,0 +1,40 @@
+run "test_required_variables" {
```
Same here.

morphis commented:
@jat-canonical while we're on it and before we official declare the plan as ready, can you sit down with @keirthana and do a review of the user facing strings?

jat-canonical committed:
chore: add copyright headers to files

jat-canonical committed:
feat(gh): use snaps instead of third party actions to setup terraform

This replaces third party GH actions with snaps and apt packages.
Since `tflint` snap is not created by an official publisher, we pin it to
a revision but that limits us to pinning to a certain architectures and
fixed GH runners.

ajanon commented on .github/workflows/build-and-test.yaml:
```diff
@@ -4,20 +4,18 @@ on:
+          sudo snap install terraform --channel latest/stable --classic
+      - name: Setup TFLint
+        run:
+         sudo snap install tflint --channel latest/stable
```
So, the tflint snap is not maintained by Canonical. Could we set it to a given revision?

Also, not sure if this has any impact, but indentation looks weird here. If it's still parsed correctly by GH actions, this is probably fine.

morphis replied:
it's maintained as it seems by Kris, who leads IS, so not a per-se no-go but I agree a trusted publisher would be good.

jat-canonical replied:
Could not locate an official one.

ajanon reviewed (changes requested):
One nitpick

jat-canonical committed:
feat(subcluster): add streaming relation for ams to relate with agent

This commit adds the relation for ams to get agent details for enabling
streaming directly from AMS.

ajanon reviewed (approved):
LGTM

jat-canonical committed:
feat(gh): move from tfsec to trivy

morphis commented on .github/actions/setup-trivy/action.yaml:
```diff
@@ -0,0 +1,102 @@
+    run: |
+      curl -L -o oras.tar.gz \
+        https://github.com/oras-project/oras/releases/download/v"${ORAS_VERSION}"/oras_"${ORAS_VERSION}"_linux_${{ inputs.arch }}.tar.gz
+      tar xf oras.tar.gz oras
```
I know this is coming from AMS and is wrong there too, but we have to actually check the sha256 checksum before moving on.

jat-canonical replied:
added

morphis reviewed (commented):
One last comment, otherwise LGTM
